File: medium _high/minimum_distance/minimum_distance.cpp
#include <iostream>
#include <vector>
#include <queue>
using ll=long long;
using namespace std;
ll mod(ll x) {return x%1000000007;}
ll n,m;
vector<vector<ll>> ship;
vector<ll> dp;
vector<ll> dist;
void bfs()
{
    queue<ll> temp;
    temp.push(0LL);
    dist[0]=0;
    ll level=0;
    while (!temp.empty())
    {
        ll sz=(ll)temp.size();
        for (ll i=0;i<sz;i++)
        {
            ll pwd=temp.front(); temp.pop();
            for (ll nxt:ship[pwd])
            {
                if (level+1LL<=dist[nxt])
                {
                    dp[nxt]=mod(dp[nxt]+dp[pwd]);
                    if (level+1!=dist[nxt])
                    {
                        dist[nxt]=level+1LL;
                        temp.push(nxt);
                    }
                }
            }
        }
        level++;
    }
    return;
}
int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cin>>n>>m;
    ship.resize(n);
    for (ll i=0;i<m;i++)
    {
        ll a,b; cin>>a>>b; a--; b--;
        ship[a].push_back(b);
        ship[b].push_back(a);
    }
    dp.assign(n,0LL); dp[0LL]=1LL;
    dist.assign(n,500000);
    bfs();
    cout<<dp[n-1LL];
    return 0;
}

File: medium _high/minimum_distance/minimum_distance.py
import sys
from collections import deque
input=sys.stdin.readline
def mod(x):
    return x%1000000007
n,m=map(int,input().split())
ship=[[] for _ in range(n+1)]
for _ in range(m):
    a,b=map(int,input().split())
    ship[a].append(b)
    ship[b].append(a)
dp=[0 for _ in range(n+1)]
dist=[500000 for _ in range(n+1)]
temp=deque()
temp.append(1)
dp[1],dist[1],level=1,0,0
while temp:
    sz=len(temp)
    for _ in range(sz):
        pwd=temp.popleft()
        for nxt in ship[pwd]:
            if level+1<dist[nxt]:
                dist[nxt]=level+1
                dp[nxt]=mod(dp[nxt]+dp[pwd])
                temp.append(nxt)
            elif level+1==dist[nxt]:
                dp[nxt]=mod(dp[nxt]+dp[pwd])
    level+=1
print(dp[n])
